### PR TITLE
ci: install perl provider

### PR DESCRIPTION
**Pull request by janlazo**
Resolve CI dependencies for https://github.com/neovim/neovim/pull/11595

### CODE REVIEW

**Comment by justinmk** [comment]
travis macOS job failed:

```
--> Working on MsgPack::Raw
Fetching http://www.cpan.org/authors/id/J/JA/JACQUESG/MsgPack-Raw-0.05.tar.gz ... OK
Configuring MsgPack-Raw-0.05 ... OK
Building MsgPack-Raw-0.05 ... FAIL
! Installing MsgPack::Raw failed. See /Users/travis/.cpanm/work/1578775862.7416/build.log for details. Retry with --force to force install it.
! Installing the dependencies failed: Module 'MsgPack::Raw' is not installed
! Bailing out the installation for Neovim-Ext-0.02.
```

**Comment by janlazo** [comment]
Is it the system perl or something else? I'll try with the perl homebrew package. Else, use perlbrew or https://github.com/tokuhirom/plenv.

File: ci/before_build.ps1 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+$ErrorActionPreference = 'Stop'
+$ProgressPreference = 'SilentlyContinue'
+
+# https://github.com/lukesampson/scoop#installation
+Invoke-Expression (New-Object System.Net.WebClient).DownloadString('https://get.scoop.sh')
+
+Set-StrictMode -Version Latest
```
Comment (janlazo) [inline ci/before_build.ps1]: If this doesn't work, scoop can't be used in the CI without disabling strict mode.

File: ci/build.ps1 [hunk 2]
```diff
@@ -57,7 +57,7 @@ if ($compiler -eq 'MINGW') {
   # in MSYS2, but we cannot build inside the MSYS2 shell.
   $cmakeGenerator = 'Ninja'
   $cmakeGeneratorArgs = '-v'
-  $mingwPackages = @('ninja', 'cmake', 'perl', 'diffutils').ForEach({
+  $mingwPackages = @('ninja', 'cmake', 'diffutils').ForEach({
```
Comment (janlazo) [inline ci/build.ps1]: Why is perl needed here again? It should safe to skip it if `C:/perl` in Appveyor is from http://strawberryperl.com/.

File: .travis.yml [hunk 3]
```diff
@@ -90,7 +91,9 @@ addons:
       - powershell
     packages:
       - ccache
+      - cpanminus
       - ninja
+      - perl
```
Comment (janlazo) [inline .travis.yml]: osx job passed so it's safe to assume that the perl provider supports recent releases only for osx.

File: ci/build.ps1 (hunk 2 above)
Comment (justinmk) [reply ci/build.ps1]: perl is needed for git, maybe other things. Though I suppose we can try this...

File: ci/before_build.ps1 (hunk 1 above)
Comment (janlazo) [reply ci/before_build.ps1]: It works.

**Comment by janlazo** [comment]
Now that perl provider can be installed in Windows,OSX,Linux, how can it be tested?

File: .travis.yml (hunk 3 above)
Comment (jacquesg) [reply .travis.yml]: It works at least all the way back to perl 5.14, but should work for perl 5.12 and 5.10 as well:

https://dev.azure.com/jacquesgermishuys/p5-Neovim-Ext/_build/results?buildId=183&view=results

Comment (janlazo) [reply .travis.yml]: Can you investigate this in your Azure pipeline by creating a new job that doesn't use perlbrew?

**Comment by justinmk** [comment]
> Now that perl provider can be installed in Windows,OSX,Linux, how can it be tested?

I assume after that, next step would be https://github.com/neovim/neovim/pull/11595 . Then perhaps someone wants to fill in `ex_perl` and friends? Here's how it was done for ruby: 0a3c0205c5b9043d371133

File: ci/build.ps1 [hunk 4]
```diff
@@ -116,6 +130,8 @@ if (-not $NoTests) {
     exit 1
   }
 
+  cpanm.bat -n Neovim::Ext
+  perl -W -e 'use Neovim::Ext; print $Neovim::Ext::VERSION'
```
Comment (justinmk) [inline ci/build.ps1]: perhaps this block should live near the npm/ruby stuff (line ~116)

**Comment by janlazo** [comment]
> could you summarize (in commit message, ideally) why scoop is a better option than e.g. powershell `Install-Product` (for node at least, if not perl).

scoop is like homebrew on Windows with the differences that it uses powershell and runs on Windows instead of using ruby and running on OSX. scoop is easy to install and manage compared to alternatives (no need for mingw/msys2/cygwin unlike pacman, works with non-admin account unlike chocolately, `Install-Product` is specific to Appveyor). scoop is regularly updated so there's less need to tinker with the package versions (no need to bump node versions anymore).

Cons would be caching and registry modification for `PATH` because of lack of integration with Appveyor. I don't know how to decouple scoop installation from the build script. It's fine for now because we're using 1 powershell script for the build and scoop is fast enough.

File: .travis.yml (hunk 3 above)
Comment (jacquesg) [reply .travis.yml]: I've tested this locally. Seems like the required perl headers are missing by default on macOS. They would need to be installed to use the system perl.

https://stackoverflow.com/questions/52682304/fatal-error-extern-h-file-not-found-while-installing-perl-modules
https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes#3035624

IMHO I think using perl from homebrew would be the simplest.

File: ci/build.ps1 (hunk 4 above)
Comment (janlazo) [reply ci/build.ps1]: It's the same install order as in `ci/install.sh`. I don't want to move it up because perl modules are often compiled. If anything, I may move it down if I need more OS packages.

File: test/functional/provider/perl_spec.lua [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  it('works', function()
+    local fname = 'Xtest-perl-hello.pl'
+    write_file(fname, [[
+      use v5.22.1;
+      use Neovim::Ext;
+      use Neovim::Ext::MsgPack::RPC;
+
+      my $session = Neovim::Ext::MsgPack::RPC::socket_session($ENV{NVIM_LISTEN_ADDRESS});
+      my $nvim = Neovim::Ext::from_session($session);
+      $nvim->command('let g:job_out = "hello"');
```
Comment (janlazo) [inline test/functional/provider/perl_spec.lua]: @jacquesg What's the simplest plugin abstraction that does this within one file (no `rplugin/perl/`)? I can't figure out how to use the Plugin class for this without creating a new file. For reference, I want to do the equivalent of what's in the nodejs_spec.lua.

Comment (janlazo) [reply test/functional/provider/perl_spec.lua]: Nevermind. I found a simple workaround.

File: test/functional/provider/perl_spec.lua [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+  end)
+
+  it('plugin works', function()
+    local fname = 'Xtest-perl-hello-plugin.pl'
+    write_file(fname, [[
+      package TestPlugin;
+      use parent qw(Neovim::Ext::Plugin);
+
+      @{TestPlugin::commands} = ();
+      @{TestPlugin::specs} = ();
```
Comment (janlazo) [inline test/functional/provider/perl_spec.lua]: @jacquesg I had to add these for perl v5.22.1 (Ubuntu Xenial) to register the plugin without warnings. Can you fix the warnings?

File: test/functional/provider/perl_spec.lua [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+      package main;
+      use v5.22.1;
+      use Neovim::Ext;
+      use Neovim::Ext::MsgPack::RPC;
+
+      my $session = Neovim::Ext::MsgPack::RPC::socket_session($ENV{NVIM_LISTEN_ADDRESS});
+      my $nvim = Neovim::Ext::from_session($session);
+      my $plugin = Neovim::Ext::Plugin::new('TestPlugin', $nvim);
+      $plugin->test_command();
```
Comment (janlazo) [inline test/functional/provider/perl_spec.lua]: This should run `TestCommand` but it needs the Host class to start the event loop. I don't know how that works.

**Comment by janlazo** [comment]
Needs a health check for perl provider and this should ready to try out before proceeding with `:perl`.

File: test/functional/provider/perl_spec.lua (hunk 7 above)
Comment (jacquesg) [reply test/functional/provider/perl_spec.lua]: A plugin should derive from ```Neovim::Ext::Plugin```

https://github.com/jacquesg/p5-Neovim-Ext/blob/master/t/rplugin/perl/TestPlugin.pm

To start the host could be as simple as [start_host()](https://github.com/jacquesg/p5-Neovim-Ext/blob/master/lib/Neovim/Ext.pm#L59)

File: test/functional/provider/perl_spec.lua (hunk 6 above)
Comment (jacquesg) [reply test/functional/provider/perl_spec.lua]: What are the warnings you are seeing?

Comment (janlazo) [reply test/functional/provider/perl_spec.lua]: `<Symbol> used only once: possible typo at <file> line <line number>`. The decorator sets these to empty arrays but perl complains that these properties aren't used.

File: test/functional/provider/perl_spec.lua (hunk 7 above)
Comment (janlazo) [reply test/functional/provider/perl_spec.lua]: Did I not do that for this test? I used `package` twice, one for the plugin and one for the session. I tried `start_host` before but it assumes `rplugin/perl` in `rtp`. I'm not testing that yet.

**Comment by janlazo** [comment]
Wonder why MSVC builds don't work or is it just 32-bit builds in general because cpan modules are compiled?

File: runtime/autoload/provider/perl.vim [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+function! provider#perl#Detect() abort
+  " use g:perl_host_prof if set or check if perl is on the path
+  let prog = exepath(get(g:, 'perl_host_prog', 'perl'))
+  if empty(prog)
+    return ''
+  endif
+
+  " if perl is available, make sure the required module is available
+  call system([prog, '-W', '-MNeovim::Ext', '-e', ''])
```
Comment (janlazo) [inline runtime/autoload/provider/perl.vim]: Used `system` because this is quick. Could use `jobstart` for https://metacpan.org/pod/local::lib, `perlbrew`, https://github.com/tokuhirom/plenv.

File: runtime/autoload/health/provider.vim [hunk 9]
```diff
@@ -623,10 +623,51 @@ function! s:check_node() abort
...
+    return
+  endif
+
+  if !executable('perl') || !executable('cpanm')
+    call health#report_warn(
+          \ '`perl` and `cpanm` must be in $PATH.',
+          \ ['Install Perl and cpanminus and verify that `perl` and `cpanm` commands work.'])
+    return
+  endif
+  let perl_v = get(split(s:system(['perl', '-W', '-e', 'print $^V']), "\n"), 0, '')
```
Comment (janlazo) [inline runtime/autoload/health/provider.vim]: Require Perl v5.6.0.

https://perldoc.perl.org/perlvar.html#%24%5eV

File: runtime/autoload/health/provider.vim [hunk 10]
```diff
@@ -623,10 +623,51 @@ function! s:check_node() abort
...
+
+  let current_perl_cmd = [host, '-W', '-MNeovim::Ext', '-e', 'print $Neovim::Ext::VERSION']
+  let current_perl = s:system(current_perl_cmd)
+  if s:shell_error
+    call health#report_error('Failed to run: '. string(current_perl_cmd),
+          \ ['Report this issue with the output of: ', string(current_perl_cmd)])
+    return
+  endif
+
+  call health#report_ok('"Neovim::Ext" cpan package is installed: '. current_perl)
```
Comment (janlazo) [inline runtime/autoload/health/provider.vim]: @jacquesg Is there a way to query cpan for latest version of Neovim::Ext?

Comment (jacquesg) [reply runtime/autoload/health/provider.vim]: You could use ```cpanm --info Neovim::Ext``` which would produce something like:

```JACQUESG/Neovim-Ext-0.02.tar.gz```

File: test/functional/provider/perl_spec.lua (hunk 7 above)
Comment (jacquesg) [reply test/functional/provider/perl_spec.lua]: The syntax to instaniate the plugin isn't quite right. It should eithe be:
```my $plugin = TestPlugin->new ($nvim)```
or
```my $plugin = new TestPlugin ($nvim)```;

https://github.com/jacquesg/p5-Neovim-Ext/blob/master/lib/Neovim/Ext/Plugin.pm#L202

It should also register the plugin like so:
https://github.com/jacquesg/p5-Neovim-Ext/blob/master/t/rplugin/perl/TestPlugin.pm#L7

**Comment by janlazo** [comment]
https://ci.appveyor.com/project/neovim/neovim/builds/30204102/job/8kiqaw62blm0mhtu
```
Building and testing IO-Async-0.75
Building IO-Async
t\00use.t .................... ok
t\01timequeue.t .............. ok
t\02os.t ..................... ok
t\03loop-magic.t ............. ok
t\04notifier.t ............... ok
t\05notifier-loop.t .......... ok
t\06notifier-mixin.t ......... ok
t\07notifier-future.t ........ ok
#   Failed test 'exceptional socket invokes on_read_ready'
#   at C:\Users\appveyor\.cpanm\work\1579357430.3792\IO-Async-0.75\blib\lib/IO/Async/LoopTests.pm line 390.
#          got: '0'
#     expected: '1'
# Looks like you failed 1 test of 22.
t\10loop-poll-io.t ........... 
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/22 subtests 
	(less 2 skipped subtests: 19 okay)
```

File: ci/install.sh [hunk 11]
```diff
@@ -50,3 +50,6 @@ else
   gcc -m32 -o "$TREE_SITTER_DIR/bin/c.so" -shared parser.c -I.
 fi
 test -f "$TREE_SITTER_DIR/bin/c.so"
+
+sudo cpanm -n Neovim::Ext || cat "$HOME/.cpanm/build.log"
```
Comment (janlazo) [inline ci/install.sh]: This fails when running `Neovim::Ext` tests because it can't download nvim v0.4.3.

Comment (jacquesg) [reply ci/install.sh]: It should only download nvim if it is not on the path, and when running tests. ```-n``` here should skip the tests?

Comment (janlazo) [reply ci/install.sh]: The issue is it can't download nvim when cpanm verified that the required Http perl modules are installed. Neovim::Ext dependencies don't fail their tests so `-n` is only for `Neovim::Ext` tests.

File: runtime/autoload/health/provider.vim (hunk 10 above)
Comment (janlazo) [reply runtime/autoload/health/provider.vim]: It works but I can't tell if that version check is for locally installed modules or from CPAN.

Comment (jacquesg) [reply runtime/autoload/health/provider.vim]: It is from CPAN, tested it on a vanilla perl.

File: ci/install.sh (hunk 11 above)
Comment (jacquesg) [reply ci/install.sh]: ```-n``` should apply to the distribution you're installing and all of its dependencies, if it doesn't, then I believe its a bug in ```App::cpanminus```. ```Neovim::Ext``` downloads neovim to verify that the API actually works. Not sure how the HTTP perl modules is in play here?

Comment (janlazo) [reply ci/install.sh]: https://github.com/jacquesg/p5-Neovim-Ext/commit/82565d72a8659ebca9196cd672e44f6c7f436b01#diff-8424792320c9483850c5842f19129c9bR89-R94

Comment (jacquesg) [reply ci/install.sh]: I don't get what your point is? Like I said above, yes, ```Neovim::Ext``` downloads nvim to run its test if its not on the ```PATH```. However, running ```cpanm``` with the ```-n```/```--notest``` switch will skip tests and the download of neovim all together.

File: ci/build.ps1 [hunk 12]
```diff
@@ -29,6 +29,26 @@ function exitIfFailed() {
...
+npm.cmd --version
+perl --version
+cpanm.bat --version
+
+cpanm.bat -n Neovim::Ext
+if ($LastExitCode -ne 0) {
+  Get-Content -Path "$env:USERPROFILE\.cpanm\build.log"
+}
+perl -W -e 'use Neovim::Ext; print $Neovim::Ext::VERSION'; exitIfFailed
+
```
Comment (justinmk) [inline ci/build.ps1]: this new block should be guarded by

    if (-not $NoTests) {

else it defeats use cases like https://github.com/neovim/neovim/pull/11654

**Review by justinmk — APPROVED** [review]
LGTM after small comment re build.ps1 https://github.com/neovim/neovim/pull/11703#discussion_r368765525

File: test/functional/provider/perl_spec.lua [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    return
+  end
+end
+
+before_each(function()
+  clear()
+end)
+
+describe('perl host', function()
+  if helpers.pending_win32(pending) then return end
```
Comment (janlazo) [inline test/functional/provider/perl_spec.lua]: https://github.com/neovim/neovim/pull/11595#discussion_r361154077
stdin polling shouldn't be a blocker for these tests on Windows since the host is not running but these tests always fail. My guess is the way I used `socket_session` to create a new session.

File: ci/before_install.sh [hunk 14]
```diff
@@ -44,7 +44,6 @@ fi
 
 source ~/.nvm/nvm.sh
 nvm install 10
```
Comment (justinmk) [inline ci/before_install.sh]: Future idea: perhaps only the "gcov" job should install ruby/python/nodejs/perl. That will make CI (1) more reliable and (2) perhaps save 30-60 seconds per job.
